### ISSUE TITLE
Idea for syntax

### ISSUE THREAD

**Issue by lhwparis** [virtualidentityag/conditional-resource-loader#1]
as discussed we have to keep our old syntax to keep compatiblilty with existing sites. for the new syntax we want to introduce i think its the best way to not rely on having the configuration in configuration.js or in the <head> of each site but as a data attribute on our components.

something like:

``` html
<div class="component" data-loader="{test: function() { return true; } , js: ['js/jquery.component.js'], css: ['css/component.css'], init: function() { $('.component').component(); } }"></div>
```

pro: 
- easier for the CMS: because thit can be render with the component and dont has to be render seperatly in the head.
- cleaner and more component based approach
- we do not need a "framework" update for new components which are only using conditionalResources

@janrembold @smohadjer @jhadenfeldt  what do you think?

**Comment by janrembold** [comment]
Looks really good! 

Just to clearify the loader syntax... 
- the init() function is the equivalent to our current scriptsDone() function. Right?
- the values of js and css should be arrays not objects

Some other questions:
- Is the new loader also jQuery-based or vanilla JS?
- Will there be a check for duplicate js/css files to exclude from loading process?
- Currently we have add() and exec() functions. I think this should be merged into a single load( config ) function. What do you think?

**Comment by timomayer** [comment]
> the init() function is the equivalent to our current scriptsDone() function. Right?

yes.

> the values of js and css should be arrays not objects

true! changed the syntax.

> Will there be a check for duplicate js/css files to exclude from loading process?

good idea! my idea was the conditionalLoader should collect all the data-loader objects. merge the js files and css files to load in one array. after thats done remove duplicates is a good addition :) after all of them are loaded all the init functions have to be executed.

> Is the new loader also jQuery-based or vanilla JS?

with the new data-attribute bases way i think we dont need the part of the loader we had in the <head> (for the new syntax) so i think its safe to rely on jquery or do you prefere a vanilla version?

> Currently we have add() and exec() functions. I think this should be merged into a single load( config ) function. What do you think?

yes thats planed and also to move this step into main.js and dont have it in the markup anymor

**Comment by janrembold** [comment]
> merge the js files and css files to load in one array. after thats done remove duplicates

This describes another procedure than a single load call. A single load would take the configured js/css files and load them immediately. You described the current way of add() and exec() above. Will there be both ways possible in the "new world"? I don't see big advantages of collecting all files.

To remove duplicate files we could push all loaded files to an internal array. So everything that comes later gets excluded from load. But that could have some tricky side effects with the ajax promises.

> i think its safe to rely on jquery or do you prefere a vanilla version?

you know i like using vanilla in any performance relevant plugins. :) but with this functionality we could reduce file size dramatically by an insignificant performance loss. So +1 for jQuery

> yes thats planed and also to move this step into main.js

i really like the idea of eliminating the main.js dependency from the conditional loader! But are there any downsides if main.js and conditionalLoader are async?

**Comment by timomayer** [comment]
> This describes another procedure than a single load call. A single load would take the configured js/css files and load them immediately. You described the current way of add() and exec() above.

i thought about that and my oppinon is that the loader gets a LOT more complex with that and we are getting in a dependency hell very quickly :/ what do you think? if we do that and one conditional plugin depends on a nother one how should we handle that and it gets more difficult if we remove duplicate scripts. when will we fire the load? i agree that it is the best way it could be done but i am not sure it is worth all the afford

> i really like the idea of eliminating the main.js dependency from the conditional loader! But are there any downsides if main.js and conditionalLoader are async?

one of the biggest benefits of the new loader will be, that main.js will not be loaded ansyc anymore but can be inside of scripts.all.js from now on. so we have a massive speed improvements for sites which are not using any conditionalResource because main.js will be executed immediately

**Comment by janrembold** [comment]
You talk about the additional complexity when scripts are used by multiple plugins and the loadings overlap asynchronously. Right? But working with promises for each script-load should make the process really easy and straightforward.

The loader singleton would need an internal "promise" object. Every load() call loads its script(s) and adds the resulting promise to the promise object. If the script was already requested it "recycles" the already loaded (or requested) promise. This "recycling" would solve the duplicate problem and the usage of promises make the script loading pretty easy. Each single load() just waits for its promises success state to fire the init() function.

And the complexity of event listeners between plugins that are not loaded already can still be solved with our failsafe events.  

The more I write about it, the more I want to start coding! :D

> one of the biggest benefits of the new loader will be, that main.js will not be loaded ansyc anymore

totally agree with you!

**Comment by smohadjer** [comment]
One disadvantage of using a data-attribute compared to configuration.js is that we lose readability, specially in cases when a component should be initialized with several parameters. We also have to first parse DOM to find all loader data which is another disadvantage. And what happens if we have ten instances of a component in a page? Are we going to repeat the data-loader attribute on every instance of the component? That is not very clean, is it?

**Comment by timomayer** [comment]
@janrembold i talk about dependencys inside the plugins. remember our project where we sit together hours to build our failsafe event handling because plugins rely on each other when it comes to correct rendering. when i keep that in mind and combine it with random initialization order (cause you never know which script is loaded first and gets initialized) i see a lot more problems coming :(

@smohadjer you are right but i see no other way. we need to be more component based with the loader and having all the loader configs in configuration.js which is a file in the framework does not help. we can not relase a new framework version everytime we want to load something conditionally and the data attribute way is better as what we do in R\* Stories at the moment where we put all the things in the <head> of each html file. thats not component based and many CMS cant render that while its easy for most CMS to render the data attributes. if a CMS is intelligent enough it can render the data attribute only on the first component of each type. if its a more dump cms, i think we ll have the attribute on each component, yes

**Comment by janrembold** [comment]
@timomayer we already have this random initialization order and solved it with the failsafe event. I don't see a reason why this loader process would make it worse. 

@smohadjer your points are absolutely valid, it's really not a clean solution. I haven't considered those facts before.

I think we could keep initialization of all plugins, directly and conditionally loaded, inside main.js. If necessary, we could still control the plugin with data attributes that overwrite default options. 

To implement this we just need to extract the test function from config object to an outer if-statement. 

``` js
// ...inside main.js
if( document.querySelector('.pluginClass') ) {
  conditionalLoader.load({ /* config and init function */ });
}
```

What do you think?

**Comment by timomayer** [comment]
@janrembold  for first: we can give it a shot and see how it works in our existing projects. i am sure for R\* it will not work. we dont have this advanced failsafe events there  and rely more on load order but its ok for me to have an option for "instant" initialization. we can try using that variant in newer projects.

i absolutly disagree with the second thing. thats not what i want to achieve with the refactoring, its more or less what we had before. the goal is that a third party agency (or even ourselfs) can build a "feature" which works in our framework without the need for a framework update. if the initialization is still in main.js this can not happen because we have to update main.js and release a new framework version to get this feature working. if its all in the component it can work without a framework update.
i agree there are cleaner ways for projects which are "whole in one" projects and not that component based as ours but we have to deal with this restrictions and some dump CMS systems in our projects and therefore i see no other solution as the one described in my first post.

**Comment by janrembold** [comment]
OK, now I got your point with the "framework update". But that means we should get rid of the whole main.js construct on the long run? 

In this case I would like to think a bit more about a better data attribute solution. For example, what about a single config file path inside the data attribute that is conditionally loaded and contains the configuration of the component. This would keep the data attribute short and readable. The downside would be two loading calls, first for the config file, second for the scripts/styles. But the filesize is extremely low and with caching enabled it wouldn't be a big deal.

**Comment by smohadjer** [comment]
That's exactly what I was thinking. This solves more or less all the problems I mentioned.

**Comment by timomayer** [comment]
@janrembold think we will always have both things, "core framework" components which will be concated into all.js and all.css files and which will be initialized in main.js. these are all the components heavily used on the site (header, grid, framework, text, tables, ....)
then we ll have components which are used rare or even only once over the whole site or just for a special event so just for a limitted time period, these are components which should not be included in the core framework to keep the core more slim and should use the conditionalLoader. (event features, searchresults, maps,...)

the size of such a config file is not the problem, its more the overhead and time each http request needs (even when its cached) one more http request which has to finish before we can start loading the resources is a downside for loading performance.
i dont agree with @smohadjer point that its not readable. yes of course its not perfect but it is still a really small snipet and data attributes were explicitly designed to handle more complex constructs like that (and libs like angular use them in exactly that way too).
if you have a syntax error and need highlighting copy the data attribute string out, debug it and put it in again.
the big benefit over an extra configuration file is that you have the information really "in the component".

one long time goal i have is to get completely rid of our styles.head.html and scripts.html file. all resource links should be defined in the component, we should rewrite useref to read all our templates, there has to be an additional flag "putResourcesInAllFiles" in the object of data-loader="", if this flag is true useref grabs the pathes from the data attribute and removes the data from the markup. if the flag is false the attribute stays. so we can use the exactly same way of resource includes for the confitionalLoader and useref and just the flag "putResourcesInAllFiles" will decide if the snipet is handled by useref (precompiled) or conditionalLoader (runtime load).

if both of you think its easier to read and maintain we can use seperate data attributes for each loader task data-loader-test, data-loader-js, data-loader-css, data-loader-init. thats more markup but makes it a little bit easier to read? what do you think?

**Comment by timomayer** [comment]
one more thing ;)
@janrembold think about your modr approach. if we handle it directly in the data attribute its easy (for the cms) to add another js file to resources when a special variant of the component is used. eg: a slider is used in the variant slider--autoslide then the cms can not only add the class slider--autoslide but also add slider.autoslide.js to the js resources. so the modular approach really plays its benefits because only the things needed get loaded. with a "static" component-config.js thats not possible or you need things like component--variant-config.js which is not very clean

**Comment by smohadjer** [comment]
@timomayer I'm not following your last comment. How does CMS know that a class like "slider--autoslide" mean that an additional resource with name slider.autoslide.js is required by the component?

**Comment by smohadjer** [comment]
I'm not totally against using data-loader attribute, but I certainly don't find them as readable or easy to debug, specially when their value becomes more complex. Your example is a simplified example and doesn't deal with path issues we encountered in R\* project which required use of a new variable for each resource to help us decide which config path to use, nor does it deal with cases when a component's init code takes html snippets as option for example. That would result in a soup of single & double quotes I don't want to be served to me! 

But it won't hurt to give it a try. If we run into problems we can always fall back on file manifest approach.

**Comment by timomayer** [comment]
> nor does it deal with cases when a component's init code takes html snippets as option for example.

thats true! but if you need a more complex init code you can even do something like load a js file called component.init.js put initialization function in there and just call it in the anonymouse "init" function ;)

>  I'm not following your last comment. How does CMS know that a class like "slider--autoslide" mean that an additional resource with name slider.autoslide.js is required by the component? 

this was more a long term vision not planned in detail yet ;) i can imagine something like we have all possible modules in a list in a data-attribute. cms parses this list and if it adds a variant which is also in that list it knows there is a js file with the same name as the variant.

to conclude that: i totally agree with both of you that its not optimal to have complex code in a data attribute, but i like to have that snipet a little bit more "dynamic" and possibly CMS generated instead of another "static" config file. i see more future benefits in the first way then in the second.

**Comment by smohadjer** [comment]
One other disadvantage of this approach, regardless of whether we embed paths in data attribute or in a manifest file is that a component is displayed in browser before its styles are available. The approach we had in R\* using conditional partial loaded components style in head, so we didn't have this problem.  Users generally don't like it when the content on a page they are reading or interacting with jumps around due to layout changes after page load.

**Comment by timomayer** [comment]
@smohadjer thats true. but what we do there is impossible for most CMS systems and will also be not possible for R\* new CMS system they start using in april so we will have to find another solution :( most CMS systems have no way to know which components are used in the page when they are rendering the <head> section. thats something we have to deal with. my approach says to load css files via conditionalLoader so via JS, i know no-js will look ugly, and if you know js is enabled you can implement a super smooth loading animation ;)

**Comment by smohadjer** [comment]
The best solution in my opinion would be that each component has its own manifest file which during build is used to add data-loader attribute to component. Isn't that something possible with gulp?

**Comment by timomayer** [comment]
interesting idea. then we have to find a way how to deal with the serve mode. think its not that performant to do that for serve too.

**Comment by smohadjer** [comment]
We don't need it for serve. Our components have only a data-config="path/to/config/file". In our code we check for data-loader attribute and if it exists we load resources using it and that's end of the story. If data-loader doesn't exist (like in our dev version) we fall back on data-config for loading component's resources.

**Comment by timomayer** [comment]
i am not sure if the step rendering the data-loader attribute via gulp only to have it written in a better readable config file is worth the afford and i am not sure if the differences between serve and build will not lead us into more problems we cant see yet.
i prefer the way to write the data-loader attribute in the data-attribute for now. when we worked a little with it and we decide a component-config and gulp rendering of the data attribute is a better way its relatively easy to update our existing code.

**Comment by janrembold** [comment]
Wow, what a long discussion. Without further ado I agree with @timomayer to try the data-loader attribute in the wild. We can adjust things later on if they are not manageable in real world environments.

**Comment by smohadjer** [comment]
I agree. I don't see why we need a test however now that we are attaching loader data directly to the component. Do we still need to test for existence of the component?

@timomayer How do we deal with scripts from different paths now? Can we use paths like 'resources/js/script.js' or ''resources-content/js/script.js'?

**Comment by timomayer** [comment]
i agree we dont need the test for most cases but maybe in some rare cases to load special polyfill files? but then we need the possibility to define tests for singe files or file bundles so maybe syntax should look like:

``` js
{resources: ['js/jquery.component.js', 'css/component.css', {test: function() { return true; }, ['js/jquery.component.polyfill.js', 'css/component.polyfill.css' ]}], init: function() { $('.component').component(); } }
```

first example is a js/css file without a test.
second one is a js/css file with a test attached to it.
changed syntx to just "resources" think there is no need to split into seperate css and js arrays.
we can use this construct not only for polyfills but also for loading modules needed for variants of the component as discussed earlier

**Comment by smohadjer** [comment]
And what about the paths?

**Comment by smohadjer** [comment]
I prefer a syntax like this:
`{
    resources: [
        {path: url, type: 'css'},
        {path: url, type: 'js'},
        {path: url, type: 'js', test: function() { return true; }}
    ]
    init: function() { $('.component').component(); }   
}`

It is more verbose, but has simpler and more consistent structure. Plus we use ajax to get scripts, but append links to html for styles, so being able to distinguish between them is useful. The only drawback is that test is defined per resource, but since we are going to rarely define tests it shouldn't matter much.

**Comment by timomayer** [comment]
thanks @smohadjer 

> The only drawback is that test is defined per resource

thats why i prefered the additional array so we can define multiple resources per test i think when we are using this tests for variants not only polyfills this would be a better apporach cause we have no redundant tests.
do we need type? cant we read that easily from the url? ending with *.js, *.css?

**Comment by timomayer** [comment]
pathes... mmh difficult question. default comes from config.resourcepath thats clear. maybe we should be able to define something like config.resourcepath2 and when this one should be used write it into the config like

{ resources: [ {path: url, base: config.resourcepath2}

**Comment by smohadjer** [comment]
So here is the final syntax in which base and test are optional. 

```
{
    resources: [
        { 
            path: 'js/jquery.component.js' 
        },
        {
            path: ['js/jquery.component.js', 'css/component.css'], 
            base: config.resourcepath           
        },
        {
            path: ['js/jquery.component.js', 'css/component.css'], 
            test: function() { return true; }, 
        }
    ], 
    init: function() { $('.component').component(); } 
}
```

**Comment by lhwparis** [comment]
Go for it. Looking forward to see it working

**Comment by smohadjer** [comment]
If we want to put functions in data-attribute, we must change syntax for functions by naming them and putting them inside ().  Then data-attribute's value can be passed to eval(). This not only makes the syntax more complex, but also means that our functions shouldn't have more than one expressions in them. There might be security concerns regarding use of eval(), but I'm not sure about that. Below is the updated syntax:

```
{
    resources: [
        { 
            path: 'js/jquery.component.js' 
        },
        {
            path: ['js/jquery.component.js', 'css/component.css'], 
            base: config.resourcepath           
        },
        {
            path: ['js/jquery.component.js', 'css/component.css'], 
            test: (function test() { return true; }), 
        }
    ], 
    init: (function init() { $('.component').component(); })
}
```

**Comment by smohadjer** [comment]
It seems in our long discussion we missed to point to biggest disadvantage of all, that we need to use eval() to parse javascript stored in data-init and data-resources. Wouldn't this open us to some liability? What if such component is used on a page where user enters personal data (e.g. a form component)? Can't some hacker abuse eval() to steal user data or tamper with it?

**Comment by timomayer** [comment]
in which way do you think this can work? eval is evil when you let the user post js in a textfield and some things like eval are not filtered out. but when we just have eval in our js code and we only execute our data attributes which a user can not manipulate i do not see any security problems

**Comment by smohadjer** [comment]
I was wondering if a hacker could tamper with data attributes on the server and change their value, but I guess that's not a concern of us, right?

**Comment by timomayer** [comment]
if a hacker has access to manipulate the data attributes he has full access to inject anything in the markup and he dont need our eval function to execute something. so i think eval is not a bad thing here

**Comment by smohadjer** [comment]
I agree. It seems all this eval fear has grown out of proportion due to not understanding what exactly the security risks are. After all I don't think anybody is stupid enough to eval a user-entered string!
